File: framework/languages/java/samples/java/TicTacToe.SessionGateway/src/main/java/systems/zlink/samples/tictactoe/sessiongateway/shared/actors/RecordingBoundSessionSendCall.java
package systems.zlink.samples.tictactoe.sessiongateway.shared.actors;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import systems.zlink.framework.actors.ZLinkBoundSessionSendCall;

public record RecordingBoundSessionSendCall(
    List<String> pushes,
    String message,
    String packetName) implements ZLinkBoundSessionSendCall {
    public RecordingBoundSessionSendCall(List<String> pushes, String message) {
        this(pushes, message, message);
    }

    @Override
    public ZLinkBoundSessionSendCall metadata(String key, String value) {
        return this;
    }

    @Override
    public ZLinkBoundSessionSendCall packetName(String packetName) {
        return new RecordingBoundSessionSendCall(pushes, message, packetName);
    }

    @Override
    public CompletionStage<Void> submitAsync() {
        pushes.add(packetName + ":" + message);
        return CompletableFuture.completedFuture(null);
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe.SessionGateway/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/sessiongateway/shared/actors/RecordingBoundSessionSendCall.kt
package systems.zlink.samples.kotlin.tictactoe.sessiongateway.shared.actors

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.framework.actors.ZLinkBoundSessionSendCall

data class RecordingBoundSessionSendCall(
    private val pushes: MutableList<String>,
    private val message: String,
    private val packetName: String,
) : ZLinkBoundSessionSendCall {
    override fun metadata(key: String, value: String): ZLinkBoundSessionSendCall = this
    override fun packetName(packetName: String): ZLinkBoundSessionSendCall =
        copy(packetName = packetName)

    override fun submitAsync(): CompletionStage<Void> {
        pushes += "$packetName:$message"
        return CompletableFuture.completedFuture(null)
    }
}
